src/rules/noUnsafeFinallyRule.ts

```diff
@@ -40,171 +39,112 @@ export class Rule extends Lint.Rules.AbstractRule {
-        const isScopeBoundary = super.isScopeBoundary(node);
-
-        return {
-            isBreakBoundary: isScopeBoundary || isLoopBlock(node) || isCaseBlock(node),
-            isContinueBoundary: isScopeBoundary || isLoopBlock(node),
-            isFinallyBlock: isFinallyBlock(node),
-            isReturnsThrowsBoundary: isScopeBoundary,
-            labels: [],
-        };
-    }
-
-    protected isScopeBoundary(node: ts.Node): boolean {
-        return super.isScopeBoundary(node) ||
-                isFinallyBlock(node) ||
-                isLoopBlock(node) ||
-                isCaseBlock(node);
-    }
-
-    private isControlFlowWithinFinallyBlock<TNode>(
-        isControlFlowBoundary: (scope: IFinallyScope, node?: TNode) => boolean, node?: TNode): boolean {
-        const scopes = this.getAllScopes();
-
-        let currentScope = this.getCurrentScope();
-        let depth = this.getCurrentDepth();
-
-        while (currentScope) {
-            if (isControlFlowBoundary(currentScope, node)) {
-                return false;
+function jumpIsLocalToFinallyBlock(jump: JumpStatement): boolean {
+    const isBreakOrContinue = jump.kind === ts.SyntaxKind.BreakStatement || jump.kind === ts.SyntaxKind.ContinueStatement;
+    const label = isBreakOrContinue ? (jump as ts.BreakStatement | ts.ContinueStatement).label : undefined;
+
+    let node: ts.Node = jump;
+    // This should only be called inside a finally block, so we'll eventually reach the TryStatement case and return.
+    while (true) {
+        const parent = node.parent!;
+        switch (parent.kind) {
+            case ts.SyntaxKind.TryStatement:
+                if ((parent as ts.TryStatement).finallyBlock === node) {
+                    return false;
+                }
+                break;
+
+            case ts.SyntaxKind.SwitchStatement:
+                if (jump.kind === ts.SyntaxKind.BreakStatement && !label) {
+                    return true;
+                }
+                break;
+
+            case ts.SyntaxKind.ForInStatement:
+            case ts.SyntaxKind.ForOfStatement:
+            case ts.SyntaxKind.ForStatement:
+            case ts.SyntaxKind.WhileStatement:
+            case ts.SyntaxKind.DoStatement:
+                if (isBreakOrContinue && !label) {
+                    return true;
+                }
+                break;
+
+            case ts.SyntaxKind.LabeledStatement: {
+                const { text } = (parent as ts.LabeledStatement).label;
+                if (label && label.text === text) {
+                    return true;
+                }
+                break;
             }
 
-            if (currentScope.isFinallyBlock) {
+            case ts.SyntaxKind.ClassDeclaration:
+            case ts.SyntaxKind.ClassExpression:
+            case ts.SyntaxKind.FunctionDeclaration:
+            case ts.SyntaxKind.FunctionExpression:
+            case ts.SyntaxKind.ArrowFunction:
                 return true;
```

Comment:
If you're climbing parents and see a function *before* you reach the try statement, that means there is a function inside the finally block. Jump statements can't leave a function, so the jump is definitely local to that finally block (because it's local to the function and the function is inside the finally block).